snappy/dpu-compress/dpu_compress.c

```diff
@@ -0,0 +1,462 @@
+
+		len -= to_write;
+		curr_index += to_write;
+		output->curr += to_write;
+		arr += to_write;
+
+		// If we are past the append window, write out current window to MRAM and start a new one
+		if (curr_index >= OUT_BUFFER_LENGTH) {
+			dbg_printf("Past EOB - writing back output %d\n", output->append_window);
+
+			mram_write(output->append_ptr, &output->buffer[output->append_window], OUT_BUFFER_LENGTH);
+			output->append_window += OUT_BUFFER_LENGTH;
+			curr_index = 0;
+		}
+	}
+}
+
+/**
+ * Copy data from the current location in the input buffer to the output buffer. 
+ * Manages the append window in the same way as the previous function.
+ *
+ * @param input: holds input buffer information
+ * @param output: holds output buffer information
+ * @param len: length of data to copy
+ */
+static void copy_output_buffer(struct in_buffer_context *input, struct out_buffer_context *output, uint32_t len)
+{
+	uint32_t curr_index = output->curr - output->append_window;
+	while (len) {
+		uint32_t to_copy = MIN(OUT_BUFFER_LENGTH - curr_index, len);
+		memcpy(&output->append_ptr[curr_index], input->ptr, to_copy);
+
+		// Advance sequential reader
+		advance_seqread(input, to_copy);
+		
+		len -= to_copy;
+		curr_index += to_copy;
+		output->curr += to_copy;
+
+		// If we are past the append window, write out current window to MRAM and start a new one
+		if (curr_index >= OUT_BUFFER_LENGTH) {
+			dbg_printf("Past EOB - writing back output %d\n", output->append_window);
+
+			mram_write(output->append_ptr, &output->buffer[output->append_window], OUT_BUFFER_LENGTH);
+			output->append_window += OUT_BUFFER_LENGTH;
+			curr_index = 0;
+		}
+	}
+}
+
+/**
+ * Get the size of the hash table needed for the size we are
+ * compressing, and reset the values in the table.
+ *
+ * @param table: pointer to the start of the hash table
+ * @param size_to_compress: size we are compressing
+ * @param table_size[out]: size of the table needed to compress size_to_compress
+ */
+static inline void get_hash_table(uint16_t *table, uint32_t size_to_compress, uint32_t *table_size)
+{
+	*table_size = 256;
+	while ((*table_size < MAX_HASH_TABLE_SIZE) && (*table_size < size_to_compress))
+		*table_size <<= 1;
+
+	memset(table, 0, *table_size * sizeof(*table));
+}
+
+/**
+ * Hash function.
+ *
+ * Any hash function will produce a valid compressed bitstream, but a good
+ * hash function reduces the number of collisions and thus yields better
+ * compression for compressible input, and more speed for incompressible
+ * input. Of course, it doesn't hurt if the hash function is reasonably fast
+ * either, as it gets called a lot.
+ *
+ * @param ptr: pointer to the value we want to hash
+ * @param shift: adjusts hash to be within table size
+ * @return Hash of four bytes stored at ptr
+ */
+static inline uint32_t hash(uint32_t ptr, int shift)
+{
+	uint32_t kmul = 0x1e35a7bd;
+	uint32_t bytes = read_uint32(ptr);
+	return (bytes * kmul) >> shift;
+}
```

Comment:
@craiig This is the hash function used by Snappy.

Reply:
lmao, ok, pretty basic! I would definitely try the built-in hash function in that case. 

Here's the built in functions UpMem defined:
https://github.com/UBC-ECE-Sasha/upmem-sdk/blob/11b9a936c8ea79cdac856307f49b6c77d097eb54/dpu-rt/src/syslib/built_ins.h

Here's the docs on the hash function, it seems to be a simple xor
https://sdk.upmem.com/2019.4.0/210_DpuHandbook.html#hash

I would try to evaluate the new hash compared to this built in hash somehow - maybe try comparing compression ratios before and after to make sure they're about the same? What's the downside to a hash collision here anyways?

Reply:
The downside of a hash collision, is you replace a 4-byte match with something that doesn't fully match. When we are searching for these matches, it will be harder to find something which results in longer literals being stored and thus worse compression.

TLDR: more collisions = worse compression

Trying out their built hash function that results in the most bits (10 bits), I can definitely see a difference. For the largest file I can compress without multiple DPUs, I go from 36% to 25% compression.

However, it does let me reduce my hash table size from 32KB to 2KB which is quite helpful. Not sure what tradeoff I should be going for here...

Reply:
However...
It does reduce the cycle count by 20%.

Reply:
Sometimes its possible to improve the distribution (i.e. reduce collisions) of simple hash functions by hashing twice with different algorithms/keys. Is that a possibility?

Reply:
I think I'm missing something - does their hash function only set 10 bits?

Reply:
Well it only XOR's 10 bits with each other. The remaining upper bits will be 0, if I understand this correctly. I did print out all the values when I was checking the function, and none are greater than 1024.

![image](https://user-images.githubusercontent.com/25714353/84842133-38083180-aff9-11ea-8ee2-ddcbf9663283.png)

@jnider Yes, we can do that. We can really do whatever we want with the hash function.